Handle parsing a PackageId that is a raw registry url into a Crate

Skgland opened the pull request:
I [noticed](https://github.com/rust-lang/rust/pull/134272#issuecomment-2551634242) that a lot of errors that I would expect to be categorized as `DependsOn(_)` would end up as `Unknown` instead.

It looks like cargo sets the `package_id` field in the json output to `registry+https://github.com/rust-lang/crates.io-index#{crate_name}@{crate_version}` which wan't handled by `Crate::try_from`.
As a result no crate would be inserted into the `deps` set and it wasn't able to differentiate between an empty `deps` set due to no errors in dependencies and an empty `deps` set because the `package_id`s field of failing deps couldn't be parsed into a `Crate`.

Skgland commented:
Looking a bit further it looks like in 1.77/1.78 cargo [switched](https://github.com/rust-lang/cargo/pull/12914) to [packeg id spec](https://doc.rust-lang.org/cargo/reference/pkgid-spec.html). 

Should I add a todo comment for implementing the parsing of the package id spec completly and cleaning up the old package id parsing? Other than that I would consider this ready for review.

pietroalbini commented on src/runner/test.rs:
```diff
@@ -233,7 +235,7 @@ fn run_cargo(
             } else if ran_out_of_space {
                 Err(e.context(FailureReason::NoSpace).into())
-            } else if !deps.is_empty() {
+            } else if let Some(deps) = deps {
```
I'm not sure about this change (setting the deps to an empty set), as the rest of Crater is not really built with support for an empty set. For example, I'm pretty sure [this code would break](https://github.com/rust-lang/crater/blob/4759ec53a64139d728afb9f4864bebe227212104/src/report/analyzer.rs#L42-L52), as it would never add the crate to the tree.

I'd be comfortable merging the first two commits, but not the third.

Skgland replied:
Would it make sense to adjust the parsing of Crate from PackageId to be infallible i.e. by having it fallback to an `Unkown(String)` variant containing the raw package id if it can't be parsed otherwise.
Then the third commit wouldn't be needed while still ensuring that the failure reason would be recorded correctly even though it might not be as nicely associated with the crate.

pietroalbini replied:
Adding a `Crate::Unknown` would then require implementing other functions of `Crate` like `to_rustwide`.

I think it's fine to miscategorize things as unknown if the pkgid changes again, as it's a clear signal something is wrong.

Skgland replied:
Ok, I have removed the third commit and instead added a new commit which adds more package id spec tests and implements more of the package id spec parsing to make those added tests pass

Skgland commented:
I have replaced the manual URL parsing code,
initially by using url::Url, 
but based on a suggestion by epage https://github.com/rust-lang/cargo/issues/15048#issuecomment-2583732687 I replaced it with [`PackageIdSpec::parse`](https://docs.rs/cargo-util-schemas/latest/cargo_util_schemas/core/struct.PackageIdSpec.html#method.parse)

pietroalbini commented on src/crates/mod.rs:
```diff
@@ -74,50 +74,134 @@ impl TryFrom<&'_ PackageId> for Crate {
+                                    .and_then(|mut path| path.next())
+                                    .unwrap_or_default()
+                                    .to_string(),
+                                name: package_id.name().to_string(),
```
Shouldn't this also be calculated from the URL? The name of the Cargo package might not match the repository name.

Skgland replied:
Oh, `name` is supposed to be the repo name not the crate name, that makes sense.

pietroalbini reviewed (changes requested):
I'm glad you found a package to parse the IDs :D

Left a small comment. Also, as before, I think we should not set the version to `"unknown"` if it's missing, but error out instead, as I don't know how the rest of Crater would handle it.

Skgland commented:
> I'm glad you found a package to parse the IDs :D
> 
> Left a small comment. Also, as before, I think we should not set the version to `"unknown"` if it's missing, but error out instead, as I don't know how the rest of Crater would handle it.

I think I missed that remark last time, I will change that then.

Skgland commented:
Ok, I have rebased to resolve merge conflicts and addressed your remarks regarding turning missing version into `unknown` and the crate-name repo-name mixup for GitHubRepo crates.

I think how I handle the `sha` field for GitRepo and GitHubRepo crates needs to be changed as well.
Currently this just puts the GitReference in there whether that's a commit has or not,
also even if it is a commit has it currently gets a `rev=` prefix added to it.

Skgland commented:
Tags, Branches and non-hex revs now result in `sha` being `None`.

pietroalbini reviewed (approved):
Thanks! :heart:

Skgland commented:
`minicrater` failed, it looks like some classifications changed, the new classification looks to be an expected improvement and this changed result just needs to be blessed, though CIs expected to actual diff is a bit hard to read as the coloring ends on linebreak, so I will need to check this locally.

Skgland commented:
Ok, I have now updated the `minicrater` expectations in the latest commit.
